hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/functional/TestHoodieBackedMetadata.java

```diff
@@ -612,214 +296,120 @@ public void testRollbackUnsyncedCommit(HoodieTableType tableType) throws Excepti
-
-    try (SparkRDDWriteClient client = new SparkRDDWriteClient(engineContext, getWriteConfig(true, true))) {
-      // Restore cannot be done until the metadata table is in sync. See HUDI-1502 for details
-      client.syncTableMetadata();
+    // trigger an upsert
+    testTable.doWriteOperation("007", UPSERT, asList("p1", "p2", "p3"), 2);
+    syncAndValidate(testTable, emptyList(), true, false, true);
 
-      // Table should sync only before the inflightActionTimestamp
-      HoodieBackedTableMetadataWriter writer =
-          (HoodieBackedTableMetadataWriter) SparkHoodieBackedTableMetadataWriter.create(hadoopConf, client.getConfig(), context);
-      assertEquals(writer.getMetadataReader().getUpdateTime().get(), beforeInflightActionTimestamp);
+    // savepoint
+    if (COPY_ON_WRITE.equals(tableType)) {
+      testTable.doSavepoint("007");
+      syncAndValidate(testTable);
+    }
 
-      // Reader should sync to all the completed instants
-      HoodieTableMetadata metadata = HoodieTableMetadata.create(context, client.getConfig().getMetadataConfig(),
-          client.getConfig().getBasePath(), FileSystemViewStorageConfig.SPILLABLE_DIR.defaultValue());
-      assertEquals(((HoodieBackedTableMetadata)metadata).getReaderTime().get(), newCommitTime);
+    // trigger delete
+    testTable.doWriteOperation("008", DELETE, emptyList(), asList("p1", "p2", "p3"), 2);
+    syncAndValidate(testTable, emptyList(), true, true, false);
 
-      // Remove the inflight instance holding back table sync
-      fs.delete(inflightCleanPath, false);
-      client.syncTableMetadata();
+    // trigger clean
+    testTable.doCleanBasedOnCommits("009", asList("001", "002"));
+    syncAndValidate(testTable, emptyList(), true, false, false);
 
-      writer =
-          (HoodieBackedTableMetadataWriter)SparkHoodieBackedTableMetadataWriter.create(hadoopConf, client.getConfig(), context);
-      assertEquals(writer.getMetadataReader().getUpdateTime().get(), newCommitTime);
+    // trigger another upsert
+    testTable.doWriteOperation("010", UPSERT, asList("p1", "p2", "p3"), 2);
+    syncAndValidate(testTable, emptyList(), true, false, false);
 
-      // Reader should sync to all the completed instants
-      metadata = HoodieTableMetadata.create(context, client.getConfig().getMetadataConfig(),
-          client.getConfig().getBasePath(), FileSystemViewStorageConfig.SPILLABLE_DIR.defaultValue());
-      assertEquals(writer.getMetadataReader().getUpdateTime().get(), newCommitTime);
-    }
+    // trigger clustering
+    testTable.doCluster("011", new HashMap<>());
+    syncAndValidate(testTable, emptyList(), true, true, false);
 
-    // Enable metadata table and ensure it is synced
+    // If there is an inflight operation, the Metadata Table is not updated beyond that operations but the
+    // in-memory merge should consider all the completed operations.
+    HoodieCommitMetadata inflightCommitMeta = testTable.doWriteOperation("007", UPSERT, emptyList(),
```

Comment:
guess we should avoid 007 at line 271. make that 008 and adjust subsequent commit times as well.